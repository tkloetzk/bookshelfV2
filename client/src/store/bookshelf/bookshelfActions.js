import * as types from './bookshelfActionTypes'
import {
  getBookshelfService,
  getGenresBookshelfService,
} from '../../services/bookshelfService'
import map from 'lodash/map'

export function getGenresSuccess(genres) {
  return {
    type: types.FETCH_GENRES_SUCCESS,
    genres,
  }
}

export function getGenres() {
  return dispatch => {
    return getGenresBookshelfService()
      .then(response => {
        const genres = map(response, genre => {
          return { label: genre, value: genre }
        })
        dispatch(getGenresSuccess(genres))
        return genres
      })
      .catch(error => {
        console.error('get genres error', error)
      })
  }
}
export function getBookshelfSuccess(bookshelf) {
  return {
    type: types.FETCH_BOOKSHELF_SUCCESS,
    bookshelf,
  }
}

export function getBookshelf() {
  return dispatch => {
    return getBookshelfService()
      .then(bookshelf => {
        dispatch(getGenres())
        dispatch(getBookshelfSuccess(bookshelf))
        return bookshelf
      })
      .catch(error => {
        console.error('bookshelf error', error)
      })
  }
}
