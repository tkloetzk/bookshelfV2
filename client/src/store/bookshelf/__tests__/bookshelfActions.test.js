import FETCH_BOOKSHELF_SUCCESS from '../bookshelfActionTypes'
import * as actions from '../bookshelfActions'
import getBookshelfService from '../../../services/bookshelfService'

jest.mock('../../../services/bookshelfService')

describe('bookshelfActions', () => {
  let bookshelf
  beforeEach(() => {
    bookshelf = [
      {
        amazonAverageRating: 4.5,
        amazonRatingsCount: 553,
        categories: ['Family & Relationships'],
        description:
          'Answers the most important parenting questions about raising children from birth through kindergarten, covering such topics as discipline, sleeping, day care, safety, independence, and feeding.',
        differences: [
          {
            currentValue: 3.9,
            key: 'amazonAverageRating',
            newValue: 4.5,
          },
        ],
        goodreadsAverageRating: 4,
        goodreadsRatingsCount: 40,
        isbn: '9781402218279',
        price: '',
        subtitle:
          'From Birth to Kindergarten, Answers to the Top 150 Questions about Raising a Young Child',
        thumbnail:
          'http://books.google.com/books/content?id=oD0omQEACAAJ&printsec=frontcover&img=1&zoom=1&source=gbs_api',
        title: 'The New Baby Answer Book',
      },
    ]
  })
  describe('getBookshelfSuccess', () => {
    it('dispatches bookshelf when successful', () => {
      const action = {
        type: FETCH_BOOKSHELF_SUCCESS,
        bookshelf,
      }
      expect(actions.getBookshelfSuccess(action.bookshelf)).toEqual(action)
    })
  })
  describe('getBookshelf', () => {
    it('should dispatch success if getBookshelfService is successful', async () => {
      const dispatch = jest.fn()

      getBookshelfService.mockResolvedValue(bookshelf)
      await actions
        .getBookshelf()(dispatch)
        .then(res => {
          expect(res).toEqual(bookshelf)
        })
      expect(dispatch.mock.calls).toMatchSnapshot()
    })
    it('should dispatch the error if getBookshelfService fails', async () => {
      const error = { id: 'error' }
      const dispatch = jest.fn()

      getBookshelfService.mockRejectedValue(error)
      await actions
        .getBookshelf()(dispatch)
        .catch(e => {
          expect(console.error).toHaveBeenCalled()
        })
    })
  })
})
