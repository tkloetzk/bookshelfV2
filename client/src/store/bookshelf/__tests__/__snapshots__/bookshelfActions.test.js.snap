// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bookshelfActions getBookshelf should dispatch success if getBookshelfService is successful 1`] = `
Array [
  Object {
    "amazonAverageRating": 4.5,
    "amazonRatingsCount": 553,
    "categories": Array [
      "Family & Relationships",
    ],
    "description": "Answers the most important parenting questions about raising children from birth through kindergarten, covering such topics as discipline, sleeping, day care, safety, independence, and feeding.",
    "differences": Array [
      Object {
        "currentValue": 3.9,
        "key": "amazonAverageRating",
        "newValue": 4.5,
      },
    ],
    "goodreadsAverageRating": 4,
    "goodreadsRatingsCount": 40,
    "isbn": "9781402218279",
    "price": "",
    "subtitle": "From Birth to Kindergarten, Answers to the Top 150 Questions about Raising a Young Child",
    "thumbnail": "http://books.google.com/books/content?id=oD0omQEACAAJ&printsec=frontcover&img=1&zoom=1&source=gbs_api",
    "title": "The New Baby Answer Book",
  },
]
`;
